### ISSUE TITLE
Deprecation of imghdr module in python 3.11

### ISSUE THREAD

**Issue by karlch** [karlch/vimiv-qt#579]
Starting from python 3.11 the `imghdr` module, which we use to check if files are images, is deprecated and will be fully removed starting from python 3.13.

I see three options moving forward:
1. Copy `imghdr.py` to `vimiv.utils`, adapt slightly, and continue using it.
2. Use a new magic library. The only one recommended in the deprecation [pep](https://peps.python.org/pep-0594/#imghdr) which seems usable for us is [filetype](https://pypi.org/project/filetype/). Simply using `filetype.is_image(...)` is not enough though, as we don't support all image formats that filetype recognizes. This is possibly also true for `imghdr`, but not as critical as these are rather obscure. filetype however also recognizes e.g. `xcf` files from gimp. So we would need to slightly update for which filetypes to check and use `filetype.match(path, matchers=CUSTOM)`.
3. Switch to mimetype-based type inference instead. I don't like this option personally and only mention it for completeness as we could just use `QtCore.QMimeDatabase` then.

I don't have a strong preference between 1. and 2. yet. Copying `imghdr` is probably straightforward and doesn't require much `vimiv` code, but puts the maintenance burden on us. `filetype` seems reasonably well packaged, the code structure is a bit weird at places (imports get overwritten all over the place, especially `filetype.types`), and could possibly allow us to remove any custom checks (if upstream supports everything). Should probably try it out and see how it goes, maybe communicating some issues upstream.

**Comment by jcjgraf** [comment]
This is unfortunate. While it introduces another dependency, to me it seems like using filetype is the way to go.

If you like, I can have a lot at it and propose something. But unfortunate not earlier than in a month.

**Comment by karlch** [comment]
I was honestly leaning the other way, given copy-pasting is nearly effortless, and the code there seems to have worked for ages :thinking:

**Comment by jcjgraf** [comment]
You are right. I had a first look at the code of imghdr, and it is indeed pretty much as simple as it gets. I just thought that since there is a well maintained package which does what we are looking for, why not using it?

**Comment by karlch** [comment]
@buzzingwire

The reason I proposed `filetype` instead of `python-magic` was the performance. This is a really dirty script that checks an image and not an image 10k times for the result and prints the time in seconds (using `timeit`):
```
# filetype
0.18984456099860836
0.15069496299838647
# magic
1.231587242000387
0.1917692560091382
# puremagic
4.894981782999821
1.919188183004735
# imghdr
0.11998687899904326
0.14208258299913723
# mime
0.19930949600529857
0.19352350500412285
```

We can see that `imghdr` is the most performant (probably thanks to limitations and simplifications, but does the job for us) and only `filetype` from the magic-based implementations comes pretty close.

Given we weren't able to just copy due to licensing, the simplicity is no longer as true, but @jcjgraf did the work already and tuned it exactly to our needs (which would have been needed in any case), I think this is an excellent effort :blush:

**Comment by buzzingwires** [comment]
That makes absolute sense, and I appreciate your taking the time to explain that. I apologize for missing this discussion, earlier. I honestly didn't expect the libmagic coupling to be so slow, but that sort of thing can be complex, aside from the obvious issues of adding the binary dependency. Rolling an implementation for the project could possibly also result in more consistent detection behavior.
